src/depackers/vorbis.c

```diff
@@ -5110,18 +5091,7 @@ int stb_vorbis_get_frame_float(stb_vorbis *f, int *channels, float ***output)
       return 0;
    }
 
-   /* Sanity check */
-   if (len > f->blocksize_1) {
-      return -1;
-   }
-
```

Comment:
1) `&len` passed to `vorbis_decode_packet`
2) `vorbis_decode_packet` first calls `vorbis_decode_initial`, receiving return values to `mode`, `p_left`, `left_end`, `p_right`, and `right_end`.
3) `len` is then passed to `vorbis_decode_packet_rest` along with the window bounds.

## `vorbis_decode_initial`

All window bounds are >= 0, <= `blocksize_1`.

### Case 1: `blockflag` == 0 or `blockflag` == 1 and the relevant `prev` or `next` flag == 1.
1) `n` = `blocksize_0` (line 3191) or `blocksize_1` (line 3186).
2) `*p_left_start` (aka `*p_left`) = 0 (line 3201).
3) `*p_left_end` (aka `*left_end`) = `window_center` = `n`/2 (line 3202).
4) `*p_right_start` (aka `*p_right`) = `window_center` = `n`/2 (line 3208).
5) `*p_right_end` (aka `*right_end`) = `n` (line 3209).
6) Since `blocksize_0` <= `blocksize_1` (line 3674), all values are less than or equal to `blocksize_1`.

### Case 2: `blockflag` == 1, `prev` == 0
1) `*p_left_start` (aka `*p_left`) = (`n` - `blocksize_0`) / 4 (line 3198) is trivially less than `blocksize_1`.
2) `*p_left_end` (aka `*left_end`) = (`n` + `blocksize_0`) / 4 (line 3199) caps out at `blocksize_1`/2.

### Case 3: `blockflag` == `, `next` == 0
1) `*p_right_start` (aka `*p_right`) = (`n`*3 - `blocksize_0`) / 4 (line 3205) is trivially less than `blocksize_1`.
2) `*p_right_end` (aka `*right_end`) = (`n`*3 + `blocksize_0`) / 4 (line 3206) caps out at `blocksize_1`.

## `vorbis_decode_packet_rest`.

`*len` is assigned in two ways. `vorbis_decode_packet_rest` returns `TRUE` after either case. `vorbis_decode_packet_rest` only returns in one other place, which is an error case where the value of `*len` won't matter.

### Normal case.

1) `*len` = `right_end` (line 3481), so <=`blocksize_1`.

### End of current stream.

If the window would cause `f->current_loc` to advance beyond `current_end`, some special handling occurs.

1) The check that triggers this: `current_end` < `f->current_loc` + (`right_end` - `left_start`).
2) If `current_end` < `f->current_loc`: `*len` is set to 0 (line 3459) + `left_start` (line 3463).
3) Otherwise: `*len` is `current_end` - `f->current_loc` (line 3461) + `left_start` (line 3463). Since (`right_end` - `left_start`) is at most `blocksize_1`, this difference shouldn't ever be larger than `blocksize_1`.
4) `*len` is immediately bounded to `right_end` (line 3464), and is thus never larger than `blocksize_1` (even if it somehow was before).
5) `*len` is always >=0, <=`blocksize_1`.

## `stb_vorbis_get_frame_float`

`*len` is always >=0, <=`blocksize_1` in cases where `vorbis_decode_packet` returns success, per above. This check (and all of the hacks that had to be added as a result of it and the other check) can now be removed because `stb_vorbis_get_frame_float` no longer returns -1 values added by libxmp-specific checks.

Check was added here: https://github.com/libxmp/libxmp/commit/a1c922ad54ecc0ba4468ea60a8db233038daddd5.
Modifications that had to be added because of negative return values from `stb_vorbis_get_frame_float`:
* https://github.com/libxmp/libxmp/commit/b4a1d365324e8171b5ed4ff91711828d0cb0d096
* https://github.com/libxmp/libxmp/commit/d815172614f0c6c5ff89f4eb06098ffcdac58349

Reply:
AFAICS, this doesn't revert https://github.com/libxmp/libxmp/commit/d815172614f0c6c5ff89f4eb06098ffcdac58349 in its entirety.

Reply:
I probably should have detailed this better, but as far as I can tell this shouldn't return -1 anymore, so both patches are no longer necessary:

* 5106: returns `error()` -> 0
* 5110: returns 0
* 5115: removed (the hardest part of this to wrap my brain around to prove :( )
* 5122: removed (`vorbis_finish_frame` never returns <0)
* 5133: returns `len`; 0 <= `len` <= `blocksize_1`

Reply:
(I wish stb_vorbis would use unsigned types for these return codes when it doesn't use negative values for errors, so this kind of ambiguity wouldn't exist in the first place.)